crates/turborepo-auth/src/token/config_token.rs

```diff
@@ -0,0 +1,135 @@
+use turbopath::AbsoluteSystemPathBuf;
+use turborepo_api_client::Client;
+
+use crate::{
+    error::Error::{FailedToFindConfigDir, FailedToReadConfigFile},
+    AuthFile, AuthToken, Error, Space, Team, TURBOREPO_AUTH_FILE_NAME, TURBOREPO_CONFIG_DIR,
+    TURBOREPO_LEGACY_AUTH_FILE_NAME,
+};
+
+#[derive(serde::Deserialize)]
+/// ConfigToken describes the legacy token format. It should only be used as a
+/// way to store the underlying token as a Token trait, and then converted to an
+/// AuthToken.
+pub struct ConfigToken {
+    pub token: String,
+}
+
+/// Attempts to read the config file for an auth token and returns the token
+/// string.
+pub fn read_config_auth() -> Result<ConfigToken, crate::Error> {
+    let config_dir = config_dir().ok_or(FailedToFindConfigDir)?;
+    let config_path = config_dir
+        .join(TURBOREPO_CONFIG_DIR)
+        .join(TURBOREPO_LEGACY_AUTH_FILE_NAME);
+
+    let body = std::fs::read_to_string(config_path).map_err(FailedToReadConfigFile)?;
+    let parsed_config: ConfigToken =
+        serde_json::from_str(&body).map_err(|e| FailedToReadConfigFile(e.into()))?;
+
+    Ok(parsed_config)
+}
+
+/// Converts our old style of token held in `config.json` into the new schema.
+///
+/// Uses the client to get information not readily available in the current
+/// token.
+pub async fn convert_to_auth_file(
+    token: &str,
+    client: &impl Client,
+) -> Result<AuthFile, crate::Error> {
+    // Get the path to the auth file.
+    let absolute_auth_path = AbsoluteSystemPathBuf::try_from(
+        config_dir()
+            .ok_or(Error::FailedToFindConfigDir)?
+            .join(TURBOREPO_CONFIG_DIR)
+            .join(TURBOREPO_AUTH_FILE_NAME),
+    )
+    .map_err(Error::PathError)?;
+
+    // Fill in auth file data.
+    let user_response = client.get_user(token).await?;
+    let teams_response = client.get_teams(token).await?;
+
+    let mut teams = Vec::new();
+    let mut af: AuthFile = AuthFile { tokens: Vec::new() };
+    // NOTE(voz): This doesn't feel great. Ideally we should async fetch all the
+    // teams and their spaces, but this should also only be invoked once in a while
+    // (until config.json doesn't have tokens anymore) so the perf hit shouldn't be
+    // a worry.
+    for team in teams_response.teams {
+        let team_id = &team.id;
+        let spaces_response = client.get_spaces(token, Some(team_id)).await?;
+        let spaces = spaces_response
+            .spaces
+            .into_iter()
+            .map(|space_data| Space { id: space_data.id })
+            .collect();
+        teams.push(Team {
+            id: team_id.to_string(),
+            spaces,
+        })
+    }
+
+    // Add the token to the list of tokens.
+    af.add_token(AuthToken {
+        token: token.to_string(),
+        api: client.base_url().to_owned(),
+        created_at: user_response.user.created_at,
+        teams,
+    });
+
+    // Write this to disk, overwriting anything that may be in `auth.json`.
```

Comment:
This is fine because if we're converting a `config.json` into `auth.json`, then there won't _be_ an `auth.json`, but it's still worth noting.